Avoid run all integration tests when net or win updated

pilou- opened the pull request:
##### SUMMARY
Don't run every integration tests when windows/networks are updated.

Fixes #25222

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
  - Bugfix Pull Request

##### COMPONENT NAME
ansible-test

pilou- committed:
Don't run all integration tests when network or windows ones are updated

pilou- committed:
templates updates don't require to run all tests

pilou- commented:
Two jobs failed:

- `freebsd/10.3-STABLE`
    ```
    2017-05-31 19:33:24 ERROR: Timeout waiting for freebsd/10.3-STABLE instance b2fe851b-1628-4ffe-9de9-77b0b7b492a5.
    ```
- `linux/ubuntu1404/2`
    ```
    2017-05-31 19:23:20 TASK [unarchive : Ensure zip is present to create test archive (apt)] **********
    2017-05-31 19:23:24 fatal: [testhost]: FAILED! => {"changed": false, "cmd": "apt-get update", "failed": true, "msg": "W: Size of file /var/lib/apt/lists/archive.ubuntu.com_ubuntu_dists_trusty-updates_restricted_binary-amd64_Packages.gz is not what the server reported 21216 23467 [...]  Hash Sum mismatch
    ```

Different successful outputs:

- `linux/centos7/1`
    ```
    2017-05-31 19:18:06 Mapping 8 changed file(s) to tests.
    2017-05-31 19:18:06 test/integration/Makefile -> all
    2017-05-31 19:18:06 WARNING: Path not categorized: test/integration/Makefile
    2017-05-31 19:18:06 test/integration/test_win_group1.yml -> all
    2017-05-31 19:18:06 WARNING: Path not categorized: test/integration/test_win_group1.yml
    2017-05-31 19:18:06 test/integration/test_win_group2.yml -> all
    2017-05-31 19:18:06 WARNING: Path not categorized: test/integration/test_win_group2.yml
    2017-05-31 19:18:06 test/integration/test_win_group3.yml -> all
    2017-05-31 19:18:06 WARNING: Path not categorized: test/integration/test_win_group3.yml
    2017-05-31 19:18:06 test/integration/windows-group1.yml -> integration: none
    2017-05-31 19:18:06 test/integration/windows-group2.yml -> integration: none
    2017-05-31 19:18:06 test/integration/windows-group3.yml -> integration: none
    2017-05-31 19:18:06 test/runner/lib/classification.py -> integration: all
    ```
- `windows/1`
    ```
    2017-05-31 19:17:47 + ansible-test windows-integration --explain --changed
    2017-05-31 19:17:47 + grep ' windows-integration: .* (targeted)$'
    2017-05-31 19:18:20 + '[' -s /tmp/windows.txt ']'
    2017-05-31 19:18:20 + echo 'Detected changes requiring integration tests specific to Windows:'
    2017-05-31 19:18:20 Detected changes requiring integration tests specific to Windows:
    2017-05-31 19:18:20 + cat /tmp/windows.txt
    2017-05-31 19:18:20 test/integration/windows-group1.yml -> windows-integration: all (targeted)
    2017-05-31 19:18:20 test/integration/windows-group2.yml -> windows-integration: all (targeted)
    2017-05-31 19:18:20 test/integration/windows-group3.yml -> windows-integration: all (targeted)
    2017-05-31 19:18:20 + '[' 1 '!=' 1 ']'
    2017-05-31 19:18:20 + echo 'Running Windows integration tests for multiple versions concurrently.'
    2017-05-31 19:18:20 Running Windows integration tests for multiple versions concurrently.
    ```
- `windows/2`
    ```
    2017-05-31 19:17:56 + ansible-test windows-integration --explain --changed
    2017-05-31 19:18:29 + '[' -s /tmp/windows.txt ']'
    2017-05-31 19:18:29 + echo 'Detected changes requiring integration tests specific to Windows:'
    2017-05-31 19:18:29 Detected changes requiring integration tests specific to Windows:
    2017-05-31 19:18:29 + cat /tmp/windows.txt
    2017-05-31 19:18:29 test/integration/windows-group1.yml -> windows-integration: all (targeted)
    2017-05-31 19:18:29 test/integration/windows-group2.yml -> windows-integration: all (targeted)
    2017-05-31 19:18:29 test/integration/windows-group3.yml -> windows-integration: all (targeted)
    2017-05-31 19:18:29 + '[' 2 '!=' 1 ']'
    2017-05-31 19:18:29 + echo 'Nothing to do, all Windows tests will run under TEST=windows/1 instead.'
    2017-05-31 19:18:29 Nothing to do, all Windows tests will run under TEST=windows/1 instead.
    2017-05-31 19:18:29 + exit 0
    ```

On windows, should not `test/runner/lib/classification.py` match `windows-integration: all` ?

mattclay commented:
I don't think there is any reason to continue supporting the legacy windows test targets in the Makefile. Those targets should be removed instead of updating ansible-test.

@nitzmahone Any objections?

pilou- commented:
@mattclay I updated the `Makefile` because I renamed the targets but I don't known how or by whom this `Makefile` is used.

Do you mean that windows targets should be removed from the `Makefile` only and that update of `classification.py` are not required ?

mattclay commented:
@pilou- The `Makefile` is the legacy method of invoking tests that predates `ansible-test`. Instead of updating the Windows related targets in the `Makefile` (and renaming their associated playbooks), they should be removed. The `ansible-test` classification code doesn't need to be updated.

pilou- committed:
ansible-test must be used instead of the legacy Makefile

pilou- commented:
@mattclay I added another commit which removes windows related targets.

With [this run](https://app.shippable.com/github/ansible/ansible/runs/23451/summary), all integration test suites were executed (integration, windows-integration and network-integration. Is that the expected result ?

mattclay commented:
@pilou- The `test_win_group*.yml` files should be removed as well, as they're only used by the windows targets in the `Makefile`.

mattclay commented on test/runner/lib/classification.py:
```diff
@@ -348,13 +348,19 @@ def _classify(self, path):
                 'network-integration': 'all',
             }
 
+        if path.startswith('test/integration/windows-'):
```
This is unnecessary as there should be no `windows-` prefixed files.

mattclay commented on test/runner/lib/classification.py:
```diff
@@ -348,13 +348,19 @@ def _classify(self, path):
+                'windows-integration': 'all',
+            }
+
+        if os.path.dirname(path) == 'test/integration' and path.endswith('.template'):
```
This is incorrect, as changes to `.template` files affect integration tests. Since changes to templates are infrequent, I wouldn't worry about adding special handing here either.

mattclay commented on test/runner/lib/classification.py:
```diff
@@ -348,13 +348,19 @@ def _classify(self, path):
             }
 
-        if path.startswith('test/integration/'):
+        if path.startswith('test/integration/network-'):
```
Handling of the `test/integration/` directory should not be removed. It's there to fall back on when there isn't a more specific match. Also, there's not much point in adding special handling for `network-` prefixed files, since there's only a single matching file which is very infrequently updated, so I'd just eliminate this change.

mattclay commented:
@pilou- Yes, that test run should run all integration tests. The only reason for that is because one of the legacy files was updated. Once they're removed that will no longer be an issue.
